### DISCUSSION TITLE
Clause segmentation

### DISCUSSION THREAD

**Discussion by 9j7axvsLuF** [explosion/spaCy#6557] [New Features & Project Ideas]
Hello,

I have been using spaCy for a while and love it. I particularly appreciate the built-in pipeline for sentence segmentation, which I use regularly to produce sentence embeddings with a Transformer language model.

However, on average, the length of sentences is inversely correlated to the quality of their embeddings. For this reason, I want to break down long/convoluted sentences into simpler clauses that could each be given their own embedding.

SpaCy does not offer any straightforward pipeline for clause segmentation. One solution would be to implement this from scratch myself using dependency parsing, but this looks like a non-trivial problem given that there are most likely edge cases that might not be adequately handle by a quick and dirty script relying on this method.

(A more sophisticated solution would be to use some kind of seq2seq model that could take a sentence like "Matt made eggs and had breakfast" and output two reconstructed sentences like "Matt made eggs" and "Matt had breakfast". But I am not aware of any pretrained model that would allow me to do that.)

Would you have any recommendation to address this issue with spaCy? Generally, I feel like it would be a great benefit to the NLP community if spaCy could handle both clause and sentence segmentation in the future.

**Comment by svlandeg** [comment]
FYI - we converted this question to a thread on the discussion board. 

We're aiming to keep the Issue tracker focused mainly on bug reports, while the new GH Discussion board can be a place where the community can come together, help eachother, discuss best practices, and so on.

**Reply by 9j7axvsLuF** [reply]
Thanks a lot!

**Comment by 9j7axvsLuF** [comment]
I found an interesting rule-based model for clause segmentation (sentence splitting) in [this paper](https://arxiv.org/pdf/2009.08560.pdf). Unfortunately the code is not shared by IBM, but here is the description of the model from Appendix A:

Given a complex sentence, the model runs the following processes once each:

## Wh Handling
Using semantic role labeling, themodel looks for a Relational Argument (R-ARG),and the Subject Argument (asserted to be the ARG preceding the R-ARG). Then, a split is made with the Relational Argument replaced by the Subject Argument.

## Conjunction Handling
The model looks for theword “and”. Using semantic role labeling, if the word following “and” is an argument (ARG), assert that “and” is followed by a sentence, and a split is made. Or, if the word following “and” is a verb (V), the model asserts the Subject Argument to be the ARG preceding the V; a split is made with “and” replaced by the Subject Argument.

## Insertion  Handling
Using dependency parsing, the model looks for a node with type participle modifier, relative  clause modifier, prepositional modifier, adjective modifier,or appositional modifier. The clause with the node as the root is extracted, prepended with the subject, and split asa new simple sentence. The rest of the original complex sentence is split as another new simple sentence.

Would anyone be able to help me implement this with spaCy?

**Reply by thomashacker** [reply]
Hey 😄 
I'm not sure if you're still interested but I'm working on something similar for `Clause Segmentation`. It's a very simple approach which only focuses on splitting compound sentences with two or more contextual statements in it.

For example: "This product has helped me recover from a back injury **but** made me nauseous"
This sentence includes two different statements about two different `conditions`

My approach was to first detect verb chunks ("has helped","made") and check if a `conjunction` lies between them.
If so, split the sentence at the respective index.

At the end I receive these two clauses:
+ "This product has helped me recover from a back injury"
+ "But made me nauseous" _(You could also delete the `conjunction` here)_

So, if you're still interested we could definitely work on an approach that might suit your usecase. 🎉

**Comment by BramVanroy** [comment]
Why not use constituency parsing instead of clauses? With your end goal in mind, that seems a reasonable alternative. IIRC the Berkeley Neural Parser (benepar) has a plugin for spaCy as well, so it should fit into your work flow easily.

**Reply by thomashacker** [reply]
That's a great plug-in and also seems to work for the newest version of `spaCy 3.1`
Here's the Link (https://spacy.io/universe/project/self-attentive-parser)
